### PR TITLE
feat: Guest token (for embedded dashboard auth)

### PR DESCRIPTION
**Pull request by suddjian**
Note: This is a PR into a feature branch, not into master. https://github.com/apache/superset/pull/17530

### SUMMARY

Adds an endpoint to create a guest token, a permission `can_grant_guest_token`, and methods in security manager to use the guest token as authn. Authorization is left out, for now.

We have decided to call this a "guest token" rather than an "embed token" because it could conceivably have future uses beyond viewing embedded dashboards.

Note that the token is not yet used for actually checking access to anything. It is only parsed to determine a "logged in" state.

### TESTING INSTRUCTIONS

tested.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/security/api.py [hunk 1]
```diff
@@ -60,3 +82,18 @@ def csrf_token(self) -> Response:
               $ref: '#/components/responses/500'
         """
         return self.response(200, result=generate_csrf())
+
+    @expose("/embedded_token/", methods=["POST"])
```
Comment (suddjian) [inline superset/security/api.py:86]: need to name this good

Comment (suddjian) [reply superset/security/api.py:86]: and add docs

File: superset/security/manager.py [hunk 2]
```diff
@@ -120,6 +121,8 @@ def __init__(self, **kwargs: Any) -> None:
 RoleModelView.edit_columns = ["name", "permissions", "user"]
 RoleModelView.related_views = []
 
+GuestToken = Dict[str, Any]
```
Comment (lilykuang) [inline superset/security/manager.py:124]: add standard jwt fields to guest token

File: superset/security/manager.py [hunk 3]
```diff
@@ -1209,3 +1219,51 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
 
         exists = db.session.query(query.exists()).scalar()
         return exists
+
+    @staticmethod
+    def create_guest_access_token(params: Dict[str, Any]):
```
Comment (lilykuang) [inline superset/security/manager.py:1224]: params: should be the same type as guest_token

File: superset/security/manager.py [hunk 4]
```diff
@@ -1209,3 +1219,51 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
...
+        self.raise_for_invalid_guest_token(token)
+        user = token["data"]["user"]
+        return GuestUser(
+            user.get("username", "guest_user"),
+            user.get("first_name", "Guest"),
+            user.get("last_name", "User")
+        )
+
+    @staticmethod
+    def decode_guest_token():
```
Comment (lilykuang) [inline superset/security/manager.py:1250]: pass in request

**Review by lilykuang — APPROVED** [review]

File: superset/security/manager.py [hunk 5]
```diff
@@ -221,6 +230,17 @@ class SupersetSecurityManager(  # pylint: disable=too-many-public-methods
...
 
+    guest_user_cls = GuestUser
+
+    def create_login_manager(self, app: Flask) -> LoginManager:
+        # pylint: disable=import-outside-toplevel
+        from superset.extensions import feature_flag_manager
+
+        lm = super().create_login_manager(app)
+        if feature_flag_manager.is_feature_enabled("EMBEDDED_SUPERSET"):
+            lm.request_loader(self.get_guest_user)
```
Comment (craig-rueda) [inline superset/security/manager.py:241]: I'm guessing this just registers another loader in the chain of "request_loaders"?

File: superset/security/manager.py [hunk 6]
```diff
@@ -1213,3 +1233,72 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
...
+        raw_token = req.cookies.get(current_app.config["GUEST_TOKEN_COOKIE_NAME"])
+        if raw_token is None:
+            return None
+
+        try:
+            token = self.parse_jwt_guest_token(raw_token)
+        except Exception as ex:  # pylint: disable=broad-except
+            # The login manager will handle sending 401s.
+            # We don't need to send a special error message.
+            logger.warning("Invalid guest token")
```
Comment (craig-rueda) [inline superset/security/manager.py:1277]: Just make one warning message:
```
logger.warning("Invalid guest token", exc_info=True)
```

File: superset/security/manager.py [hunk 7]
```diff
@@ -1213,3 +1233,72 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
...
+            logger.warning(ex)
+            return None
+        else:
+            user = token["user"]
```
Comment (craig-rueda) [inline superset/security/manager.py:1281]: Can you add a method to delegate this to? I'm thinking that folks may want to override the claims, etc. and that will make it easier.

File: superset/security/manager.py [hunk 8]
```diff
@@ -1213,3 +1233,72 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
...
+        now = self._get_current_epoch_time()
+        exp = now + (current_app.config["GUEST_TOKEN_JWT_EXP_SECONDS"] * 1000)
+        claims = {
+            "user": user,
+            "resources": resources,
+            # standard jwt claims:
+            "iat": now,  # issued at
+            "exp": exp,  # expiration time
+        }
+        token = jwt.encode(claims, secret, algorithm="HS256")
```
Comment (craig-rueda) [inline superset/security/manager.py:1256]: Can you make the `algo` configurable?

**Review by craig-rueda — APPROVED** [review]
Looks good - left a few structural suggestions

File: superset/security/manager.py (hunk 7 above)
Comment (suddjian) [reply superset/security/manager.py:1281]: I'm not sure how exactly you're saying folks might want to override things. They could add extra claims or whatever already by overriding `parse_jwt_guest_token` and calling `super()`.

We could also delegate the [creation of claims](https://github.com/apache/superset/pull/17517/files/f67c8a86efb41f522e0710e45b3fcf439d5b0335#diff-bcb3868b9e29aea5f6e87ec641120dae35a86d732501f488a7a6dbdeb450f5c1R1249-R1255) to allow putting extra stuff into the token.

File: superset/security/manager.py (hunk 5 above)
Comment (suddjian) [reply superset/security/manager.py:241]: There can only be one request loader set, no chain. But currently Superset doesn't set one. If there is a user session active, the loader doesn't get called, and if the loader returns None it's treated as if no session was found.

I know that in Shell we have our own request loader, and I've already made a branch where that is modified to be friendly with this change.

File: superset/security/manager.py (hunk 7 above)
Comment (craig-rueda) [reply superset/security/manager.py:1281]: If you override, you'll need to clobber the contents of this method. It's likely that extra claims will need to be added, etc. I'm sure this works as-is - just future-proofing

Comment (suddjian) [reply superset/security/manager.py:1281]: I think I get what you mean. I'm going to modify the GuestUser class to take the token instead, and then extra claims can be added by subclassing GuestUser. That should cover those concerns.

**Pull request merged**